### ISSUE TITLE
Remap lang: simplify renaming fields

### ISSUE THREAD

**Issue by JeanMertz** [timberio/vector#5150]
Currently, to rename a field, you'd do:

```rust
.bar = .foo
del(.foo)
```

Renaming multiple fields, you'd do:

```rust
.bar = .foo
.qux = .baz
del(.foo, .baz)
```

Especially as the list of fields grows, and you're overwriting some existing fields, but not all, it can become tedious to remember which fields you have to delete at the end.

Since renaming fields is a very common use-case, it makes sense to optimize the UX for this.

---

There are many ways we can tackle this, but here are a few that I think are most-likely to be useful. Feel free to expand on this list if you have an alternative you think is worth considering.

## 1. `del` returns value of deleted field

By having `del` return the value of the deleted field (it currently always returns `null`), you could do:

```rust
.foo = del(.bar)
.qux = del(.baz)
```

The downside of this proposal is that `del` currently allows deleting up to 16 fields at once:

```rust
del(.foo, .bar, .baz, .qux, ...)
```

If we return the value of the deleted field, we'd have to do one of four things:

1. only accept a single field per `del` call
2. always return `null` (similar to current behaviour) if more than one field is passed to `del`
3. return the first or last value in the list of provided fields, ignoring the rest
4. provide a new `del_all` (or similarly named) function to allow deleting multiple fields without getting a return value.

I'm inclined to favour either (1) or (2), but neither is great.

## 2. add a new `rename` function

We could add an explicit `rename` function, which would probably have a signature like:

```rust
fn rename(path = Path, name = String)
```

This would allow you to do:

```rust
rename(.bar, "foo")
rename(.baz, "qux")
```

The reason the second argument is a string and not a path, is because we have to prevent something like:

```rust
rename(.foo.bar.baz, .qux)
```

Which wouldn't work, since this would imply we're _moving_ data from one field to another.

## 3. add a new `move` function

Similar to `rename`, but make it explicit that you're moving values around, not renaming fields:

```rust
fn move(from = Path, to = Path)
```

```rust
move(.baz, .foo) 			// value from `.baz` moved to `.foo`, field `.baz` deleted
move(.foo.bar.baz, .qux) 	// value from `.foo.bar.baz` moved to `.qux`, field `.baz` deleted
```

This function could have an optional third parameter `compact`, which would also remove `.bar` and `.foo` if they did not have any fields any more. However, we could forego this third parameter, favouring an explicit `compact` function (#4430).

This method would also work for indexing into arrays:

```rust
move(.foo[2], .foo[1])
move(.foo.bar[1], .qux)
move(.foo.bar, .baz[4])
```

The one downside to the above example is this one:

```rust
move(.foo[0], .foo[1])
```

Since we're moving data within the array itself, _and_ we're actually moving a lower value to a higher index, we'd have to do one of three things, all having their own downsides:

1. index `0` is assigned the `null` value, index `1` is assigned the old value of index `0`
2. index `1` is assigned the value of `0`, index `0` is removed, making index `1` the new `0`th index
3. a runtime error is returned, indicating that this is not supported

A similar problem occurs if you assign a value to a higher index than is currently present in the array, although I consider this at least a bit less unsurprising:

```rust
move(.foo, .bar[100])
```

This would either:

1. generate 100 `null` values if the array doesn't exist yet (or any number between the highest index and 100)
2. return a runtime error

_(note that this is already possible using `.foo[99] = "bar"`, but this is currently not working as expected, I'm re-doing path querying/assignment logic, which will fix this)_

## 4. add custom language syntax for move operations

I'm adding this one since it's an obvious alternative to the function-driven approach of the first three proposals.

However, I don't really have a concrete proposal on _what_ syntax we'd use for this, so I'll just leave this here in case anyone else does.

As an example (this could be interpreted as a logical bit shift operator, which is why I don't have a great example to give):

```rust
.foo << .bar // move `.bar` into `.foo`
```

**Comment by binarylogic** [comment]
I prefer 1 to start. It's the simplest and most straight-forward.

**Comment by jszwedko** [comment]
I'm not sure what the best option is, but some thoughts:

For 1, could we allow multiple assignment? Like `.baz,, boop = del(.foo, .bar)`. Otherwise, I like the idea of restricting `del()` to a single field rather than introducing different return behavior based on whether one or more fields are passed.

For 4, I might suggest `.foo <- .bar` rather than `<<`.

**Comment by binarylogic** [comment]
I'm leaning towards 1 with the following changes:

1. Change `del` to only accept a single field. We can solve mass operations with a higher level syntax like comprehensions (ex: `[path in .prefix_* { del(path) }]`
2. Also change any other functions that take multiple arguments to only take a single one.
3. Have `del(...)` return the value.

Therefore we end up with:

```
.new_field = del(.old_field)
```

**Comment by JeanMertz** [comment]
@binarylogic rightfully pointed out that we probably want to solve "iterating" as a higher level language construct at some point.

One such example would be a `walk` function/operation to traverse maps/arrays and operate on each individual field:

```rust
walk(.foo, del(.)) // delete all fields below the `.foo` path
```

Another is simply allowing you to iterate over a list of expressions and act on them:

```rust
[.foo, .bar, .baz] -> del(.)
```

_(I'm making syntax up as I go here, please ignore)_

But essentially, we should change the `del` function to only work on one field and remove the `only_fields` function. Then we'd later solve that problem in a more generic way in the language, instead of per-function.

**Comment by bruceg** [comment]
I agree with @binarylogic option 1 is the easiest path forward in the short term, particularly given the complications in the other options.

> If we return the value of the deleted field, we'd have to do one of four things:

5. Return a list containing all the deleted fields

**Comment by FungusHumungus** [comment]
It would be nice to make sure that with option 1. we don't clone the whole field in the process. I think at the minute this would happen.

A dedicated function that purely renames the key whilst keeping the field data in place would be more efficient.

**Comment by JeanMertz** [comment]
This was solved in #5633.
